vmngclient/api/repository_api.py

```diff
@@ -0,0 +1,105 @@
+import logging
+import os
+from enum import Enum
+from typing import Dict, List, Union
+
+from attr import define, field
+
+from vmngclient.dataclasses import Device
+from vmngclient.session import Session
+from vmngclient.utils.creation_tools import FIELD_NAME, create_dataclass, get_logger_name
+
+logger = logging.getLogger(get_logger_name(__name__))
+
+
+class DeviceCategory(Enum):
+    VSMART = "controllers"
+    VBOND = "controllers"
+    VEDGE = "vedges"
+    CEDGE = "vedges"
+    VMANAGE = "controllers"
+
+
+@define
+class DeviceSoftwareRepository:
+    installed_versions: List[str] = field(default=None)
+    available_versions: List[str] = field(default=None, metadata={FIELD_NAME: "availableVersions"})
+    current_version: str = field(default=None, metadata={FIELD_NAME: "version"})
+    default_version: str = field(default=None, metadata={FIELD_NAME: "defaultVersion"})
+    device_id: str = field(default=None, metadata={FIELD_NAME: "uuid"})
+
+
+class RepositoryAPI:
+    """
+    API methods to prepare data for most of software actions
+    """
+
+    def __init__(
+        self,
+        session: Session,
+        devices: List[Device],
+        device_category: DeviceCategory,
+    ):
+
+        self.session = session
+        self.devices = []
+        self.device_category = device_category
+
+        for dev in devices:
+            dev_dict = dict()
+            dev_dict["deviceId"] = dev.uuid
+            dev_dict["deviceIP"] = dev.id
+            self.devices.append(dev_dict)
+
+    def get_image_version(self, software_image: str) -> Union[str, None]:
+        """
+        Method to get proper software image version
+
+        Args:
+            software_image (str): path to software image
+
+        Returns:
+            Union[str, None]: image version or None
+        """
+        url = "/dataservice/device/action/software/images?imageType=software"
+        image_name = os.path.basename(software_image)
+        software_images = self.session.get_data(url)
+        for img in software_images:
+            if image_name in img["availableFiles"]:
+                image_version = img["versionName"]
+                return image_version
+        logger.error(f"Software image {image_name} is not in available images")
+        return None
+
+    def create_devices_versions_repository(self) -> Dict[str, DeviceSoftwareRepository]:
```

Comment:
it grabs data from endpoint usding GET and serializes it. Maybe rename it to get_devices_software_repository_dict(...)